From 4a73984f287b2f85adb14223361bcd38ef45632e Mon Sep 17 00:00:00 2001
From: Claire Girka <claire@sitedethib.com>
Date: Tue, 26 Jul 2022 11:13:45 +0200
Subject: [PATCH] [draft] winebus: store container sysfs path from udev
 backend.

Store the sysfs path of container USB devices detected by the udev backend
so a container ID can be assigned.
---
 dlls/winebus.sys/bus_udev.c | 34 ++++++++++++++++++++++++++++++++++
 dlls/winebus.sys/unixlib.h  |  1 +
 2 files changed, 35 insertions(+)

diff --git a/dlls/winebus.sys/bus_udev.c b/dlls/winebus.sys/bus_udev.c
index 332e79e..c3e3241 100644
--- a/dlls/winebus.sys/bus_udev.c
+++ b/dlls/winebus.sys/bus_udev.c
@@ -1496,6 +1496,36 @@ static const struct hid_device_vtbl lnxev_device_vtbl =
 };
 #endif /* HAS_PROPER_INPUT_HEADER */
 
+static void get_device_container_syspath(const char *syspath, struct device_desc *desc)
+{
+    char path[MAX_PATH], buffer[10], *p;
+
+    if (strlen(syspath) + strlen("/removable") + 1 > MAX_PATH)
+        return;
+
+    strcpy(path, syspath);
+
+    while ((p = strrchr(path, '/'))) {
+        FILE *f;
+
+        strcpy(p, "/removable");
+        f = fopen(path, "r");
+        *p = 0;
+
+        if (f) {
+            if (fgets(buffer, 10, f) && strcmp(buffer, "fixed") != 0) {
+                /* It's a potentially removable device, so treat it as a container */
+                fclose(f);
+                break;
+            }
+            fclose(f);
+        }
+    }
+
+    if (p && (p - path) > 12)
+        lstrcpynA(desc->container_syspath, path, sizeof(desc->container_syspath));
+}
+
 static void get_device_subsystem_info(struct udev_device *dev, const char *subsystem, const char *devtype,
                                       struct device_desc *desc, int *bus)
 {
@@ -1560,6 +1590,9 @@ static void get_device_subsystem_info(struct udev_device *dev, const char *subsy
 
         if ((tmp = udev_device_get_sysattr_value(parent, "serial")))
             ntdll_umbstowcs(tmp, strlen(tmp) + 1, desc->serialnumber, ARRAY_SIZE(desc->serialnumber));
+
+        if (!desc->container_syspath[0] && (tmp = udev_device_get_syspath(dev)))
+            get_device_container_syspath(tmp, desc);
     }
 }
 
diff --git a/dlls/winebus.sys/unixlib.h b/dlls/winebus.sys/unixlib.h
index 02e7a1c..c6c74a3 100644
--- a/dlls/winebus.sys/unixlib.h
+++ b/dlls/winebus.sys/unixlib.h
@@ -45,6 +45,7 @@ struct device_desc
     WCHAR manufacturer[MAX_PATH];
     WCHAR product[MAX_PATH];
     WCHAR serialnumber[MAX_PATH];
+    char container_syspath[MAX_PATH];
 };
 
 struct sdl_bus_options

